### ISSUE TITLE
Question about transaction rollback if import failure

### ISSUE THREAD

**Issue by martinmart92** [b2ihealthcare/snow-owl#439]
After an import failure (cf https://github.com/b2ihealthcare/snow-owl/issues/438), is it rolled-back like in a transactional database, or is there some manual cleaning to do to remove part of the data imported successfully in Elasticsearch?

**Comment by cmark** [comment]
Hi @martinmart92,

Sorry for my late reply.

tl;dr;
You can remove that data manually, but unfortunately, it's not that easy to select the changes to revert. Feel free to re-execute the commit request (be it a simple change or an import).


There is a minimal transactionality built into Snow Owl 7.x, something like what Git does when you push changes to a remote repository, but that's by nature is not like in an RDBMS (also there is no traditional transactionality in Elasticsearch). 
Commit requests can happen parallel on different branches, but only one can be applied in a given time to a given branch because we lock that branch in the given repository.
If something goes wrong during a commit on a branch, there is no rollback, because of two reasons. 1. Elasticsearch does not have a rollback (nor a transaction) feature. We would need to manually remove the partially committed changes from the index, which are basically new documents (new revisions) and revised revisions (basically marking a revision revised from a branch's perspective). 
2. Since the revisions feature is not a mission-critical scenario and something more like how a team of people author content (be it code, book or terminology) it is easier to let the content be erroneous and with the help of tools like validation and verification help the team resolves the errors and make the content consistent again (like an eventual consistency model).

This has been discussed a few times internally and the conclusion was always the same: it would be nice to have a rollback feature and with enough commit failures reported by the community we might implement it in the future, but for now, this model works really well and so far there were no issues with it. 

I hope this answers your questions, feel free to reopen this issue if you have more.

Cheers,
Mark
